dll/win32/shell32/dialogs/filetypes.cpp

```diff
@@ -1462,129 +1636,166 @@ EditTypeDlgProc(HWND hwndDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
+FileTypesDlg_SetGroupboxText(HWND hwndDlg, LPCWSTR Assoc)
 {
-    HWND hListView;
-    DWORD dwIndex = 0;
-    WCHAR szName[50];
-    WCHAR szFile[100];
-    DWORD dwName;
-    LVITEMW lvItem;
-    INT iItem = 0;
-    HIMAGELIST himlLarge, himlSmall;
-
-    // create imagelists
-    himlLarge = ImageList_Create(GetSystemMetrics(SM_CXICON), GetSystemMetrics(SM_CYICON),
-                                 ILC_COLOR32 | ILC_MASK, 256, 20);
-    himlSmall = ImageList_Create(GetSystemMetrics(SM_CXSMICON), GetSystemMetrics(SM_CYSMICON),
-                                 ILC_COLOR32 | ILC_MASK, 256, 20);
-
-    // set imagelists to listview.
-    hListView = GetDlgItem(hwndDlg, IDC_FILETYPES_LISTVIEW);
-    ListView_SetImageList(hListView, himlLarge, LVSIL_NORMAL);
-    ListView_SetImageList(hListView, himlSmall, LVSIL_SMALL);
-
-    FileTypesDlg_InitListView(hwndDlg, hListView);
-
-    szFile[0] = 0;
-    if (!LoadStringW(shell32_hInstance, IDS_FILE_EXT_TYPE, szFile, _countof(szFile)))
-    {
-        // default to english
-        wcscpy(szFile, L"%s File");
-    }
-    szFile[(_countof(szFile)) - 1] = 0;
+    CStringW buf;
+    buf.Format(IDS_FILE_DETAILS, Assoc);
+    SetDlgItemTextW(hwndDlg, IDC_FILETYPES_DETAILS_GROUPBOX, buf.GetString());
+}
 
-    dwName = _countof(szName);
+static void
+FileTypesDlg_Refresh(HWND hwndDlg, HWND hListView, PFILE_TYPE_GLOBALS pG)
+{
+    ListView_DeleteAllItems(hListView);
+    ImageList_RemoveAll(pG->himlSmall);
+    InitializeDefaultIcons(pG);
+    FileTypesDlg_SetGroupboxText(hwndDlg, L"");
+    SetDlgItemTextW(hwndDlg, IDC_FILETYPES_DESCRIPTION, L"");
+    SetDlgItemTextW(hwndDlg, IDC_FILETYPES_APPNAME, L"");
+    EnableWindow(GetDlgItem(hwndDlg, IDC_FILETYPES_DELETE), FALSE);
+    RedrawWindow(hwndDlg, NULL, NULL, RDW_ALLCHILDREN | RDW_INVALIDATE | RDW_UPDATENOW);
+#if DBG
+    DWORD statTickStart = GetTickCount();
+#endif
 
+    INT iItem = 0;
+    WCHAR szName[ASSOC_CCHMAX];
+    DWORD dwName = _countof(szName);
+    DWORD dwIndex = 0;
+    SendMessage(hListView, WM_SETREDRAW, FALSE, 0);
     while (RegEnumKeyExW(HKEY_CLASSES_ROOT, dwIndex++, szName, &dwName,
                          NULL, NULL, NULL, NULL) == ERROR_SUCCESS)
     {
-        if (FileTypesDlg_InsertToLV(hListView, szName, iItem, szFile))
+        if (FileTypesDlg_InsertToLV(hListView, szName, iItem, pG))
             ++iItem;
         dwName = _countof(szName);
     }
+    FileTypesDlg_Sort(pG, hListView);
+    SendMessage(hListView, WM_SETREDRAW, TRUE, 0);
+    RedrawWindow(hListView, NULL, NULL, RDW_ALLCHILDREN | RDW_ERASE | RDW_FRAME | RDW_INVALIDATE);
+
+#if DBG
+    WCHAR dbgbuf[200];
+    wsprintfW(dbgbuf, L"%u (%ums)", iItem, GetTickCount() - statTickStart);
+    RECT r;
+    GetWindowRect(hwndDlg, &r);
+    HWND hDbg = CreateWindowExW(WS_EX_TRANSPARENT, L"STATIC", dbgbuf, WS_CHILD | WS_DISABLED | SS_RIGHT |
+                                WS_VISIBLE, 0, 0, r.right - r.left, pG->IconSize, hwndDlg, 0, NULL, NULL);
+    struct Hide {
```

Comment:
lmao, so this is the trick to allow sub-routine definition inside a function? :joy:

Reply:
Yeah, it's the same as a lambda only done by hand.

It's to keep this debugging code all in one place.